Can't determine Project to build. Expected 1 .csproj or .fsproj but found 3 when running "func start"

The `func start` command used to work flawlessly but it started showing the subject error message. Further info:

- .NET SDK 8.0.101
- `func --version` reveals version 4.0.5455
- There is only one .csproj file in the function app's folder, but the function app is a part of a C# solution
- I launched VS Code to create and initialize a new minimal azure function app. The `func start` works.

To reproduce the issue:

- create new Azure Function App project using VSCode with HttpTrigger
- new project works fine and function app runs locally
- save project as workspace. SLN file appears in project directory.
- project is now unusable.

> /usr/local/share/dotnet/sdk/8.0.100/NuGet.targets(399,5): error MSB3202: The project file "/Users/alistair/dev/test/TestFunction/obj/Debug/net8.0/WorkerExtensions/WorkerExtensions.csproj" was not found. [/Users/alistair/dev/test/TestFunction/TestFunction.sln]

Existing projects and workspaces are fine. It's just workspaces created with this version of VSCode:

> Version: 1.90.0
> Commit: 89de5a8d4d6205e5b11647eb6a74844ca23d2573
> Date: 2024-06-04T19:34:44.157Z
> Electron: 29.4.0
> ElectronBuildId: 9593362
> Chromium: 122.0.6261.156
> Node.js: 20.9.0
> V8: 12.2.281.27-electron.0
> OS: Darwin x64 22.6.0
> 

**### Deleting the sln file resolves the issue but VSCode recreates it when opening the workspace and destroys the project.**

--- a/src/Azure.Functions.Cli/Common/FileSystemHelpers.cs
+++ b/src/Azure.Functions.Cli/Common/FileSystemHelpers.cs
@@ -170,18 +170,20 @@ namespace Azure.Functions.Cli.Common
                 }
             }
 
-            if (searchOption == SearchOption.AllDirectories)
+            if (searchOption == SearchOption.TopDirectoryOnly)
             {
-                foreach (var directory in Instance.Directory.GetDirectories(directoryPath, "*", SearchOption.TopDirectoryOnly))
+                yield break;
+            }
+                
+            foreach (var directory in Instance.Directory.GetDirectories(directoryPath, "*", SearchOption.TopDirectoryOnly))
+            {
+                var directoryName = Path.GetFileName(directory);
+                if (excludedDirectories == null ||
+                    !excludedDirectories.Any(d => d.Equals(directoryName, StringComparison.OrdinalIgnoreCase)))
                 {
-                    var directoryName = Path.GetFileName(directory);
-                    if (excludedDirectories == null ||
-                        !excludedDirectories.Any(d => d.Equals(directoryName, StringComparison.OrdinalIgnoreCase)))
+                    foreach (var file in GetFiles(directory, excludedDirectories, excludedFiles, searchPattern, searchOption))
                     {
-                        foreach (var file in GetFiles(directory, excludedDirectories, excludedFiles, searchPattern, searchOption))
-                        {
-                            yield return file;
-                        }
+                        yield return file;
                     }
                 }
             }

--- a/src/Azure.Functions.Cli/Common/FileSystemHelpers.cs
+++ b/src/Azure.Functions.Cli/Common/FileSystemHelpers.cs
@@ -158,7 +158,7 @@ namespace Azure.Functions.Cli.Common
             }
         }
 
-        internal static IEnumerable<string> GetFiles(string directoryPath, IEnumerable<string> excludedDirectories = null, IEnumerable<string> excludedFiles = null, string searchPattern = "*")
+        internal static IEnumerable<string> GetFiles(string directoryPath, IEnumerable<string> excludedDirectories = null, IEnumerable<string> excludedFiles = null, string searchPattern = "*", SearchOption searchOption = SearchOption.AllDirectories)
         {
             foreach (var file in Instance.Directory.GetFiles(directoryPath, searchPattern, SearchOption.TopDirectoryOnly))
             {
@@ -170,15 +170,18 @@ namespace Azure.Functions.Cli.Common
                 }
             }
 
-            foreach (var directory in Instance.Directory.GetDirectories(directoryPath, "*", SearchOption.TopDirectoryOnly))
+            if (searchOption == SearchOption.AllDirectories)
             {
-                var directoryName = Path.GetFileName(directory);
-                if (excludedDirectories == null ||
-                    !excludedDirectories.Any(d => d.Equals(directoryName, StringComparison.OrdinalIgnoreCase)))
+                foreach (var directory in Instance.Directory.GetDirectories(directoryPath, "*", SearchOption.TopDirectoryOnly))
                 {
-                    foreach (var file in GetFiles(directory, excludedDirectories, excludedFiles, searchPattern))
+                    var directoryName = Path.GetFileName(directory);
+                    if (excludedDirectories == null ||
+                        !excludedDirectories.Any(d => d.Equals(directoryName, StringComparison.OrdinalIgnoreCase)))
                     {
-                        yield return file;
+                        foreach (var file in GetFiles(directory, excludedDirectories, excludedFiles, searchPattern, searchOption))
+                        {
+                            yield return file;
+                        }
                     }
                 }
             }

--- a/src/Azure.Functions.Cli/Helpers/DotnetHelpers.cs
+++ b/src/Azure.Functions.Cli/Helpers/DotnetHelpers.cs
@@ -155,8 +155,9 @@ namespace Azure.Functions.Cli.Helpers
         public static bool CanDotnetBuild()
         {
             EnsureDotnet();
-            var csProjFiles = FileSystemHelpers.GetFiles(Environment.CurrentDirectory, searchPattern: "*.csproj").ToList();
-            var fsProjFiles = FileSystemHelpers.GetFiles(Environment.CurrentDirectory, searchPattern: "*.fsproj").ToList();
+            // dotnet build will only search for .csproj files within the current directory (when no .csproj file is passed), so we limit our search to that directory only
+            var csProjFiles = FileSystemHelpers.GetFiles(Environment.CurrentDirectory, searchPattern: "*.csproj", searchOption: SearchOption.TopDirectoryOnly).ToList();
+            var fsProjFiles = FileSystemHelpers.GetFiles(Environment.CurrentDirectory, searchPattern: "*.fsproj", searchOption: SearchOption.TopDirectoryOnly).ToList();
             // If the project name is extensions only then is extensions.csproj a valid csproj file
             if (!Path.GetFileName(Environment.CurrentDirectory).Equals("extensions"))
             {

